sockeye/prepare_data.py

```diff
@@ -69,19 +69,37 @@ def prepare_data(args: argparse.Namespace):
     logger.info("Adjusting maximum length to reserve space for a BOS/EOS marker. New maximum length: (%d, %d)",
                 max_seq_len_source, max_seq_len_target)
 
+    # 1. Split input into shards and randomly assign data to shards
+    num_sents = 0
+    with utils.smart_open(source_paths[0], mode='rb') as infile:
+        for _ in infile:
+            num_sents += 1
+    num_shards = data_io.get_num_shards(num_sents, samples_per_shard, minimum_num_shards)
+    logger.info("%d samples will be split into %d shard(s) (requested samples/shard=%d, min_num_shards=%d)."
+                % (num_sents, num_shards, samples_per_shard, minimum_num_shards))
+    shards = data_io.create_shards(source_fnames=source_paths,
+                                   target_fnames=target_paths,
+                                   num_shards=num_shards,
+                                   output_prefix=output_folder)
+    shard_source_paths, shard_target_paths = [paths for paths in zip(*shards)]
+
+    # Process shards in parallel using max_processes process    
+    pool = multiprocessing.pool.Pool(processes=args.max_processes)
```

Comment:
I agree, it would get quite complicated...
What was the main reason for having a variant for the single process? Was it because execution was slower with a pool of a single worker process?

Reply:
I think mainly to not have the overhead of `pool` when there is only a single process/single shard. If we wanted to support single process without pool the easiest would probably to define a dummy `Pool` that implements all relevant pool methods but uses Python `map` etc. instead.